src/factorio_randovania_mod/factorio_formats.py

```diff
@@ -0,0 +1,147 @@
+        self.message = message
+        self.flagbuildnone = True
+        self.error = error
+
+    def _parse(self, stream, context, path):
+        message = construct.evaluate(self.message, context)
+        raise self.error(f"Error field was activated during parsing with error {message}", path=path)
+
+    def _build(self, obj, stream, context, path):
+        message = construct.evaluate(self.message, context)
+        raise self.error(f"Error field was activated during building with error {message}", path=path)
+
+    def _sizeof(self, context, path):
+        raise construct.SizeofError("Error does not have size, because it interrupts parsing and building", path=path)
+
+
+PropertyTreeType = construct.Enum(
+    construct.Int8ul,
+    none=0,
+    bool=1,
+    number=2,
+    string=3,
+    list=4,
+    dictionary=5,
+    signedinteger=6,
+    unsignedinteger=7,
+)
+
+SpaceOptimizedUInt = construct.Select(
+    construct.FocusedSeq(
+        "val",
+        "val" / construct.Int8ul,
+        "check" / construct.Check(construct.this.val < 255),
+    ),
+    construct.FocusedSeq(
+        "val",
+        "magic" / construct.Const(255, construct.Int8ul),
+        "val" / construct.Int32ul,
+    ),
+)
+
+PropertyString = construct.FocusedSeq(
+    "value",
+    empty=construct.Rebuild(construct.Flag, lambda ctx: ctx == ""),
+    value=construct.If(construct.this.empty == False, construct.PascalString(SpaceOptimizedUInt, "utf-8")),
+)
+
+_property_type_value = {
+    "none": construct.Pass,
+    "bool": construct.Flag,
+    "number": construct.Double,
+    "string": PropertyString,
+    "signedinteger": construct.Int64sl,
+    "unsignedinteger": construct.Int64ul,
+}
+
+
+def _python_value_to_tree_type(value: typing.Any) -> str:
+    if value is None:
+        return "none"
+    if isinstance(value, bool):
+        return "bool"
+    if isinstance(value, str):
+        return "string"
+    if isinstance(value, float):
+        return "number"
+    if isinstance(value, list):
+        return "list"
+    if isinstance(value, dict):
+        return "dictionary"
+    if isinstance(value, int):
+        if value > 2**63:
+            return "unsignedinteger"
+        else:
+            return "signedinteger"
+    raise construct.ConstructError(f"Unsupported type: {value}")
+
+
+PropertyTree = construct.FocusedSeq(
+    "value",
+    "type"
+    / construct.Rebuild(
+        PropertyTreeType,
+        lambda ctx: _python_value_to_tree_type(ctx.value),
+    ),
+    construct.Const(False, construct.Flag),
+    "value"
+    / construct.Switch(
+        construct.this.type,
+        _property_type_value,
+        ErrorWithMessage(lambda ctx: f"Unknown type: {ctx.type}"),
+    ),
+)
+
+_property_type_value["list"] = construct.PrefixedArray(
+    construct.Int32ul,
+    construct.FocusedSeq(
+        "value",
+        construct.Const("", PropertyString),
+        "value" / PropertyTree,
+    ),
+)
+_property_type_value["dictionary"] = DictAdapter(
```

Comment:
weird that these two are separated so far from the initial dict definition

Reply:
`_python_value_to_tree_type` came later and made it weird, but there's a circular dependency where these two need Property Type and that needs this dict defined.